### PR TITLE
Round beta / delta_tau to get number of slices in DQMC

### PR DESCRIPTION
**Pull request by MasonProtter**
I'm working on making a DQMC version of https://github.com/crstnbr/MonteCarlo.jl/blob/master/example/Ising2D.jl. Currently, the documented interface for `DQMC` says that I specify `beta` and `delta_tau` and then it calculates `slices` from their quotient and slices must be an integer. It's annoying that one has to fine tune `delta_tau` so that `beta / delta_tau` is an integer. I propose that this should just be rounded, especially because due to floating point error, choosing `delta_tau = beta / 500` does not always mean `beta / delta_tau` can be converted to an integer.

### CODE REVIEW

**Comment by carstenbauer** [comment]
Thanks for bringing this up. This should be discussed and handled in one way or another. In my private dqmc code I can either specify `beta` or the number of `slices`, to avoid rounding error situations.

**Comment by carstenbauer** [comment]
I'm wondering whether we should at least produce a warning when `round`ing. After all `round` does more than just correcting floating point numerical errors.

Example: `beta/delta_tau = 3.321/0.1`. Arguably not very realistic but technically allowed. It is not clear to me what the users wants in this case. An error/warning would probably be appropriate. Just setting `slices` to 33 with no warning/error could potentially be misleading.

Maybe we should even look at the difference `beta/delta_tau - round(beta/delta_tau)` and check that it is "floating point error small". If yes, `round`, otherwise, warn/error.

Any thoughts?

**Comment by MasonProtter** [comment]
> I'm wondering whether we should at least produce a warning when rounding. After all round does more than just correcting floating point numerical errors.

Yeah, this branch is actually just a local change I made so that a script I'm testing could run. There's definitely a lot more we can do to make it clear to a user what is going on. 

Putting your above comments together, it seems that a nice solution would be to have three separate methods concerning `beta`, `delta_tau` and `slices`.

- User specifies `beta` and `slices`. We then compute `delta_tau`.
- User specifies `delta_tau` and `slices`. We then compute `beta`. 
- User specifies `beta` and `delta_tau`. We then compute `slices`. With this method, we could do something like `@assert beta/delta_tau ≈ round(beta/delta_tau)`, or we could emit a warning, or we could ask that the user also provide their own rounding function. That way they can choose between say `floor`, `ceil` and `round` or make a function like `myround(x) = x ≈ round(x) ? round(x) : error(" ... ")`

**Comment by MasonProtter** [comment]
Okay, so it took some really ugly conditionals and whatnot since we can't dispatch on keyword arguments, but I got everything working. The interface I went with for now is

- User specifies beta and slices. We then compute delta_tau.
- User specifies delta_tau and slices. We then compute beta.
- User specifies beta and delta_tau. We then compute slices with `round(beta/delta_tau)`, and emit a warning if `!(round(beta/delta_tau) !≈ beta/delta_tau)`.

Thoughts / comments?

File: src/flavors/DQMC/DQMC.jl [hunk 1]
```diff
@@ -15,21 +15,59 @@ end
...
+                        global_rate::Int    = 5,
+                        thermalization::Int = 100, 
+                        sweeps::Int         = 100,
+                        all_checks::Bool    = true, 
+                        safe_mult::Int      = 10,
+                        measure_rate::Int   = 10,
+                        kwargs...)
+    nt = (;kwargs...)
+    keys(nt) == (:beta)
+    nt = (;beta=nt.beta, delta_tau=0.1)
```
Comment (ffreyer) [inline src/flavors/DQMC/DQMC.jl]: Did you mean to have

```julia
keys(nt) == (beta,) && ( nt = (;beta=nt.beta, delta_tau=0.1))
```

here?

Comment (MasonProtter) [reply src/flavors/DQMC/DQMC.jl]: Yep!

**Comment by carstenbauer** [comment]
Looks good to me!

**Comment by MasonProtter** [comment]
Not sure why the Travis CI isn't running.

File: src/flavors/DQMC/DQMC.jl [hunk 2]
```diff
@@ -37,8 +37,7 @@ function DQMCParameters(;global_moves::Bool = false,
                         measure_rate::Int   = 10,
                         kwargs...)
     nt = (;kwargs...)
-    keys(nt) == (:beta)
-    nt = (;beta=nt.beta, delta_tau=0.1)
+    keys(nt) == (:beta) && nt = (;beta=nt.beta, delta_tau=0.1)
```
Comment (carstenbauer) [inline src/flavors/DQMC/DQMC.jl]: The assignment needs to be in paratheses.

**Review by carstenbauer — CHANGES_REQUESTED** [review]

File: src/flavors/DQMC/DQMC.jl (hunk 2 above)
Comment (MasonProtter) [reply src/flavors/DQMC/DQMC.jl]: Done

**Comment by carstenbauer** [comment]
Gazillions of github CI jobs :)

(They will probably fail)

**Comment by carstenbauer** [comment]
Can you fix the tests? It would also be good to have a few tests for the different keyword options.

**Comment by MasonProtter** [comment]
These tests pass locally on my machine (2017 MacBook Pro i5 on MacOS Catalina). Not sure what's wrong. Possibly related to BLAS / LAPACK non-determinism.

**Comment by carstenbauer** [comment]
Yeah, probably something like that. This exact test has also errored for one of the master test runs and magically worked again in the following run. Can you open a separate issue so that we don't forget to fix this?

I'll ignore the test failure here and merge.

**Comment by carstenbauer** [comment]
I opened an issue myself: https://github.com/crstnbr/MonteCarlo.jl/issues/50

**Pull request merged**